### ISSUE TITLE
Custom Authentication Flow

### ISSUE THREAD

**Issue by kshaaban-** [driftyco/ionic-platform-web-client#86]
This is not a bug/defect, but rather a request for enhancement to the current flow for custom user authentication.

I'm working on implementing custom authentication for my app, but wasn't really digging being forced to show the InAppBrowser to endusers even briefly after a successful login; and also didn't want it to stay open to communicate an error in authentication.

What I did was modifying [this part of the Auth Class](https://github.com/driftyco/ionic-platform-web-client/blob/master/src/auth/auth.js#L91) to

```javascript
//...
                                                     // Hide the InAppBrowser --------v
      var tempBrowser = window.cordova.InAppBrowser.open(loc, '_blank', 'location=no,hidden=yes');
      tempBrowser.addEventListener('loadstart', function (data) {
        if (data.url.slice(0, 20) === 'http://auth.ionic.io') {
          var queryString = data.url.split('#')[0].split('?')[1];
          var paramParts = queryString.split('&');
          var params = {};
          for (var i = 0; i < paramParts.length; i++) {
            var part = paramParts[i].split('=');
            params[part[0]] = part[1];
          }
          storeToken(authOptions, params.token);
          tempBrowser.close();
          deferred.resolve(true);
        } else if (data.url === 'http://some/incorrect/credentials') {
          // Added this else if block where I redirect from my server in case
          // of invalid login credentials I will probably pass some parameters
          // along with this URL to notify endusers through the rejected promise
          tempBrowser.close();
          deferred.reject(false);
        }
      });
//...
```

What do you guys think of this implementation? and what are the chances of integrating something along these once made customizable? I'd be more than happy to contribute with your guidance.

**Comment by ericb** [comment]
We've already gotten similar requests and it's definitely on the list of things to look into. Ideally inappbrowser would only be a requirement if you *need* it to be (for redirect-style authentication). I'm going to keep this open as an enhancement/question so that we can update with what we change in the future.

Thanks !

**Comment by astik** [comment]
@kshaaban- , you say "This is not a bug/defect", but in the actual state of the ionic.io client, when credentials are wrong, the user experience is very bad, don't you think ?
The fact that in case of wrong credentials, being able to get back to the form and display a nice error message would be way better.
The solution you gave in your post works perfectly for us =)
@ericb as it does not seem to be a priority, what is the ionic's correct usage ?

**Comment by ericb** [comment]
@astik What do you mean exactly?

**Comment by astik** [comment]
Here is the user workflow :

1. user submit a form with username and password into ionic application
2. controller triggers "$ionicAuth.login('custom', authOptions, data)"
3. the login function triggers the opening of an inAppBrowser as it is a custom provider
4. ionic server is called to retrieve the custom backend url
5. inAppBrowser is redirected to the custom backend url

Credentials are correct :

6a. credentials are ok, a redirect url is issued
7a. user is redirected to ionic server to finish the authentication (custom user is created into ionic.io)
8a. inAppBrower is redirected to auth.ionic.com which is a fake serve, it is identified as the end of the login workflow, inAppBrower is closed.
9a. as login is ok, the associated JS promise is resolved successfully
10a. user informations are retrieved
11a. user is authenticated, informations are loaded in ionic app, they lived happily ever after

Credentials are incorrect :

6b.credentials are KO, a 401 status code is issued
7b. inAppBrowser display the 401 message and does not get closed
8b. user is sad as the screen display a technical error message and not the nice form he started his authentication worflow with

So to summarize, what I found bad in this design is the fact that in case of error, user does not get back to the form. Also, you seem very cool in your first message, so that, I though you had insight for a way to get around this problem =)

**Comment by ericb** [comment]
@astik Shouldn't the custom auth server be handling the error before redirecting back to us? Even so I still see the need for us to give a way for an error to be sent back to us which can ultimately be used with the platform client (ie. the inappbrowser would be closed, and the app code could handle an error sent back via the server).

**Comment by astik** [comment]
@ericb handling the error on the custom auth server would be an option. That would get the following modification to the previous workflow :

6b->6c. credentials are KO, a page saying that the credentials are bad, in a human way, is displayed
7c. as user needs to access the form to fix his credentials, the error page triggers a redirect after few seconds (javascript or meta markup in the error page, ...) or after a user action on a button or a link
8c. inAppBrowser is redirected to auth.ionic.com and gets closed (like in 8a)
9c. credentials were bad but the closing of inAppBrowser resolved the promise successfully, javascript will try to retrieve user informations (processes chained through promise resolution)
10c. as user is not authenticated, user informations retrieval finishes with a 401 error on ionic server.

In that scenario, this is still not very convenient.

Error handling on custom auth backend is not handy ... the backend could be shared with multiple ionic application, the error page may need a lot of design for each application. That leads to : for each new application deployed using this custom auth server, a new error page needs to be deployed (depending on the app_id or through a new specific endpoint).

**Comment by kshaaban-** [comment]
@astik I'm glad my workaround worked well for you! 

I'm pretty sure this will be improved in the future though.

**Comment by ericb** [comment]
Yup, I agree. We will definitely be addressing this with future updates to the Auth components.

**Comment by mos-ie** [comment]
Hey guys, any update on this?

**Comment by arendjantetteroo** [comment]
Also interested in this. We already had a fine authentication workflow, so now we only call the ionic auth one after i know the user has the correct credentials, so it first authenticates against our own login backend and then calls another ionic auth call with the inappbrowser, just displaying a white page for a 1-2 seconds. It's annoying and ugly but works. 

But now with trying to get wkwebview working, this white page is no longer auto closed but needs to be clicked on done to close it. 

A way to have an invisble inappbrowser (or another route that does not need the inappbrowser at all) would be appreciated. Is there another way to do this that i missed?

**Comment by dwieeb** [comment]
I have an option to hide the InAppBrowser window in the new Cloud Client. I'm considering this issue closed, as our new documentation website (next week!) will feature a migration guide from `ionic-platform-web-client` and details on how to hide the InAppBrowser window.
